File: packages/core/src/schema.ts
/**
 * Gemeinsames Export-Schema von CLI und GUI (Version 2).
 *
 * Vertrag: `wiki/Export-Schema.md`. Die GUI erzeugt dasselbe JSON in
 * `tauri-app/src-tauri/src/schema.rs`; beide Seiten werden zusammen geändert.
 * Feldnamen bleiben snake_case wie in den Rust-Strukturen, damit ein Export
 * ohne Umbenennung zwischen den Oberflächen ausgetauscht werden kann.
 *
 * Gegenüber Version 1 entfallen `kompakt_ausgelassen` und `abbruch_grund`:
 * Limits brechen jetzt **vor** der Verarbeitung ab (Fail Loud), es gibt kein
 * Teilergebnis mehr, das einen Abbruchgrund erklären müsste.
 */
import { KontextDatei } from './datei';

/** Version des Schemas: bei jeder inkompatiblen Änderung erhöhen. */
export const SCHEMA_VERSION = 2;

/** Eine Datei im Export. */
export interface SchemaDatei {
  relativer_pfad: string;
  sprache: string;
  zeilen: number;
  zeichen: number;
  inhalt: string;
}

/**
 * Kopf- und Zählerdaten eines Exports.
 *
 * Auch intern in snake_case: dieselben Objekte gehen sowohl in den
 * Markdown-Header als auch in das JSON, so dass es keine zweite Schreibweise
 * gibt, die auseinanderlaufen könnte.
 */
export interface SchemaMeta {
  titel: string;
  zeitstempel: string;
  uebersprungen: number;
}

/** Vollständiger Exportinhalt. */
export interface SchemaKontext extends SchemaMeta {
  schemaVersion: number;
  gesamt_dateien: number;
  gesamt_zeilen: number;
  gesamt_zeichen: number;
  dateien: SchemaDatei[];
}

/**
 * Zeitstempel im gemeinsamen Format (lokale Zeit, Sekunden).
 *
 * Bewusst nicht `toLocaleString`: das Schema soll in jedem Gebietsschema
 * dieselbe Form haben wie `chrono` in der GUI.
 */
export function zeitstempelJetzt(): string {
  const jetzt = new Date();
  const zwei = (wert: number) => String(wert).padStart(2, '0');
  const datum = `${jetzt.getFullYear()}-${zwei(jetzt.getMonth() + 1)}-${zwei(jetzt.getDate())}`;
  const uhrzeit = `${zwei(jetzt.getHours())}:${zwei(jetzt.getMinutes())}:${zwei(jetzt.getSeconds())}`;
  return `${datum} ${uhrzeit}`;
}

/** Baut das Export-Objekt; die Reihenfolge der Dateien bleibt erhalten. */
export function kontextObjekt(
  meta: SchemaMeta,
  dateien: KontextDatei[]
): SchemaKontext {
  return {
    schemaVersion: SCHEMA_VERSION,
    titel: meta.titel,
    zeitstempel: meta.zeitstempel,
    gesamt_dateien: dateien.length,
    gesamt_zeilen: dateien.reduce((summe, datei) => summe + datei.zeilen, 0),
    gesamt_zeichen: dateien.reduce((summe, datei) => summe + datei.zeichen, 0),
    uebersprungen: meta.uebersprungen,
    dateien: dateien.map(datei => ({
      relativer_pfad: datei.relativerPfad,
      sprache: datei.sprache,
      zeilen: datei.zeilen,
      zeichen: datei.zeichen,
      inhalt: datei.inhalt,
    })),
  };
}

/** JSON-Ausgabe (zwei Leerzeichen Einrückung, wie `to_string_pretty` in Rust). */
export function kontextAlsJson(
  meta: SchemaMeta,
  dateien: KontextDatei[]
): string {
  return JSON.stringify(kontextObjekt(meta, dateien), null, 2);
}

File: tauri-app/src-tauri/src/schema.rs
//! Gemeinsames Export-Schema (Version 2) des JSON-Exports.
//!
//! Vertrag: `wiki/Export-Schema.md`. Gegenstück im Core:
//! `packages/core/src/schema.ts`.
//! Beide Seiten müssen zusammen geändert werden; Feldnamen und Reihenfolge
//! sind dort beschrieben. Bewusst ohne `absoluter_pfad`: der Export wandert in
//! geteilte Kontexte und darf keine lokalen Pfade verraten.
//!
//! Gegenüber Version 1 entfallen `kompakt_ausgelassen` und `abbruch_grund`:
//! Limits brechen jetzt **vor** der Verarbeitung ab (Fail Loud), es gibt kein
//! Teilergebnis mehr, das einen Abbruchgrund erklären müsste.

use crate::scan::ScanErgebnis;
use serde::Serialize;

/// Version des Schemas: bei jeder inkompatiblen Änderung erhöhen.
pub const SCHEMA_VERSION: u32 = 2;

/// Eine Datei im Export.
#[derive(Debug, Serialize)]
pub struct SchemaDatei<'a> {
    pub relativer_pfad: &'a str,
    pub sprache: &'a str,
    pub zeilen: usize,
    pub zeichen: usize,
    pub inhalt: &'a str,
}

/// Vollständiger Exportinhalt.
#[derive(Debug, Serialize)]
pub struct SchemaKontext<'a> {
    #[serde(rename = "schemaVersion")]
    pub schema_version: u32,
    pub titel: &'a str,
    pub zeitstempel: &'a str,
    pub gesamt_dateien: usize,
    pub gesamt_zeilen: usize,
    pub gesamt_zeichen: usize,
    pub uebersprungen: usize,
    pub dateien: Vec<SchemaDatei<'a>>,
}

/// Baut das Export-Objekt aus einem Scan-Ergebnis; die Reihenfolge der Dateien
/// bleibt erhalten.
pub fn aus_scan(scan: &ScanErgebnis) -> SchemaKontext<'_> {
    SchemaKontext {
        schema_version: SCHEMA_VERSION,
        titel: &scan.titel,
        zeitstempel: &scan.zeitstempel,
        gesamt_dateien: scan.dateien.len(),
        gesamt_zeilen: scan.gesamt_zeilen,
        gesamt_zeichen: scan.gesamt_zeichen,
        uebersprungen: scan.uebersprungen,
        dateien: scan
            .dateien
            .iter()
            .map(|datei| SchemaDatei {
                relativer_pfad: &datei.relativer_pfad,
                sprache: &datei.sprache,
                zeilen: datei.zeilen,
                zeichen: datei.zeichen,
                inhalt: &datei.inhalt,
            })
            .collect(),
    }
}
